### ISSUE TITLE
doc: unclear what router.param does

### ISSUE THREAD

**Issue by rattrayalex** [alexmingoia/koa-router#124]
Hi, 
I just don't get what `router.param` really does from reading the docs.  Does it intercept any route with `:some_param` in the definition and then... do things to it? How?

I think providing a more complete example might be helpful, including example url's and return values. 

If this is just me being thick, kindly ignore =) but most other things on the page made sense.

**Comment by rattrayalex** [comment]
Perhaps this would be more clear:

```js
router
  .param('userId', function *(id, next) {
    this.user = yield db.users.getById(id);
    if (!this.user) return this.status = 404;
    yield next;
  })
  .get('/users/:userId', function *(next) {
    this.body = this.user;
  })
  .get('/users/:userId/friends', function *(next) {
    this.body = yield this.user.getFriends();
  })
// /users/3 => {"id": 3, "name": "Alex"}
// /users/3/friends => [{"id": 4, "name": "TJ"}]
```

**Comment by rattrayalex** [comment]
While I'm here, a few other questions from the docs: 

- `router.register` is just a low-level `router.all`, `router.get`, etc, right?
- what is the callback in `router.all`? What is the difference between that and a middleware?
- would `userRouter.prefix('users')` be preferred to `router.use('users', require('userRouter')` or similar? if so, why?

**Comment by rattrayalex** [comment]
One way I see of doing this:
```js
// in app.js
app.use(require('userRouter').routes());
app.use(require('taskRouter').routes());
app.use(require('someOtherRouter').routes());
```
```js
// in eg; taskRouter.js
export router
  .prefix('/tasks')
  .param('id', function *(id, next) {
    this.task = yield tasks.get(id);
  })
  .get('/', function *(next) {
    this.body = yield tasks.getAll();
  })
  .get('/:id', function *(next) {
    this.body = this.task;
  })
```
Or, perhaps:
```js
// in app.js
let taskRouter = new Router({prefix: '/tasks'})
  .param('id', require('./controllers/tasks').processParam)
  .get('/', require('./controllers/tasks').list)
  .get('/:id', require('./controllers/tasks').show)
  .routes()

app.use(new Router({prefix: '/users'})  // or app.use('/users', new Router() ?
  .param('id', require('./controllers/users').processParam)
  .get('/', require('./controllers/users').list)
  .get('/:id', require('./controllers/users').show)
  .use(taskRouter) // as child router - is this possible? eg; /users/7/tasks
  .routes()
)
app.use(taskRouter)
```

Is there one way that's more "right" than the others? Are there relative advantages/drawbacks to each?

**Comment by alexmingoia** [comment]
I've updated the `router.param()` example with yours to be more clear.

To answer your questions:
- `router.register()` is the low-level method for registering routes. It was not meant to be included in the docs and is part of the internal API. I have since updated the docs to exclude it and mark it as private.
- `router.all()` is just like `router.get()` or `router.post()` except that it registers the callback with all HTTP methods. It's only real purpose is as an internal method to perform redirects for `router.redirect()`.
- The newly released `5.0.0` has support for nesting routers. So you can do `router1.get('/users', router2.routes())`. You can also accomplish the same thing by setting a prefix in the router options.

**Comment by rattrayalex** [comment]
Wow, awesome! Thanks so much @alexmingoia !! The new support for nested routers looks great in particular.
